cf-oscore/src/main/java/org/eclipse/californium/oscore/HashMapCtxDB.java

```diff
@@ -59,20 +64,62 @@
 	public HashMapCtxDB() {
 
 		this.tokenMap = new HashMap<>();
-		this.ridMap = new HashMap<>();
+		this.contextMap = new HashMap<>();
 		this.uriMap = new HashMap<>();
 		this.seqMap = new HashMap<>();
 		this.allTokens = new ArrayList<Token>();
 	}
 
+	/**
+	 * Retrieve context using RID and ID Context. If the provided ID Context is
+	 * null a result will be returned if there is only one unique context for
+	 * that RID.
+	 */
 	@Override
-	public synchronized OSCoreCtx getContext(byte[] rid) {
-		if (rid != null) {
-			return ridMap.get(new ByteId(rid));
-		} else {
+	public synchronized OSCoreCtx getContext(byte[] rid, byte[] IDContext) throws CoapOSException {
+		// Do not allow a null RID
+		if (rid == null) {
 			LOGGER.error(ErrorDescriptions.BYTE_ARRAY_NULL);
 			throw new NullPointerException(ErrorDescriptions.BYTE_ARRAY_NULL);
 		}
+
+		HashMap<ByteId, OSCoreCtx> matchingRidMap = contextMap.get(new ByteId(rid));
+
+		// No matching RID found at all
+		if (matchingRidMap == null) {
+			return null;
+		}
+
+		// If a RID was found get the specific context
+		if (IDContext == null) {
+			// If retrieving using only RID, there must be only 1 match maximum
+			if (matchingRidMap.size() > 1) {
+				throw new CoapOSException(ErrorDescriptions.CONTEXT_NOT_FOUND_IDCONTEXT, ResponseCode.UNAUTHORIZED);
+			} else {
+				// If only one entry return it
+				Map.Entry<ByteId, OSCoreCtx> first = matchingRidMap.entrySet().iterator().next();
+				return first.getValue();
+			}
+
+		} else {
+			// If retrieving using both RID and ID Context
+			return matchingRidMap.get(new ByteId(IDContext));
+		}
+	}
+
+	/**
+	 * Retrieve context using only RID when it is certain it is unique.
```

Comment:
I would prefer a exception, if that "certain" is broken.

Reply:
Yes, better to check and throw an exception. I have now added this in commit 53c1507badbae05cbd12830f8e750701eab8f3bb